study/src/test/java/transaction/stage1/Stage1Test.java

```diff
@@ -173,64 +139,37 @@ void noneRepeatable() throws SQLException {
      *   Read phenomena | Phantom reads
      * Isolation level  |
      * -----------------|--------------
-     * Read Uncommitted |
-     * Read Committed   |
-     * Repeatable Read  |
-     * Serializable     |
+     * Read Uncommitted | +
+     * Read Committed   | +
+     * Repeatable Read  | +
+     * Serializable     | -
      */
     @Test
     void phantomReading() throws SQLException {
```

Comment:
MySQL의 경우 일반적인 RDMS와 달리 특수한 기능으로 인해 Phantom Read의 발생을 막을 수 있습니다!
어떤 방식으로 Phantom Read를 방지하는지, 그럼에도 Phantom Read가 발생할 수 있는 상황이 있는지 궁금하네요!

Reply:
> To prevent phantoms, InnoDB uses an algorithm called **next-key locking** that combines index-row locking with gap locking. InnoDB performs row-level locking in such a way that when it searches or scans a table index, it sets shared or **exclusive locks on the index records** it encounters. Thus, the row-level locks are actually index-record locks. In addition, a next-key lock on an index record also affects the **“gap” before the index record.**

MySQL 에서는 next key locking 이라는 알고리즘을 사용한다고 합니다.
1. 해당하는 row 에 락을 걸고
2. row 사이(gap)에 락을 거는 방식입니다.

하지만 next key locking은 인덱스에 대해서 작동되기 때문에, 인덱스가 없는 테이블에서는 팬텀 리드가 발생할 수 있을 것 같습니다🤔

ref. https://dev.mysql.com/doc/refman/8.0/en/innodb-next-key-locking.html

Reply:
락에대해 잘 학습해주셨네요!
`next key locking은 인덱스에 대해서 작동되기 때문에, 인덱스가 없는 테이블에서는 팬텀 리드가 발생할 수 있을 것 같습니다`
해당 답변에 대해서 저는 select for update 시 인덱스가 없다면 테이블 전체에 락을 거는 것으로 알고있어요!
그래서 인덱스의 유무보다는 
1. select 쿼리 실행해서 범위 조회
2. 다른 트랜잭션에서 범위 사이에 데이터 삽입후 커밋
3. select for update 로 다시 범위 조회 시 언두 로그가 아닌 실제 테이블을 조회하므로 새로 커밋된 데이터까지 조회

이런 일반적이지 않은 상황에서 일어날 수 있다는 것을 말씀드리고 싶었어요!